core/commands/filestore.go

```diff
@@ -0,0 +1,946 @@
+func (f *fixPath) Read(res []byte) (int, error) { return 0, io.EOF }
+func (f *fixPath) FileName() string             { return f.orig.FileName() }
+func (f *fixPath) FullPath() string             { return f.orig.FullPath() }
+func (f *fixPath) Close() error                 { return f.orig.Close() }
+
+func (f *fixPath) NextFile() (files.File, error) {
+	f0, _ := f.orig.NextFile()
+	if f0 == nil {
+		return nil, io.EOF
+	}
+	if len(f.paths) == 0 {
+		return nil, errors.New("len(req.Files()) < len(req.Arguments())")
+	}
+	path := f.paths[0]
+	f.paths = f.paths[:1]
+	if f0.IsDirectory() {
+		return nil, errors.New("Online, directory add not supported, try '-S'")
+	} else {
+		f, err := os.Open(path)
+		if err != nil {
+			return nil, err
+		}
+		stat, err := f.Stat()
+		if err != nil {
+			return nil, err
+		}
+		return &dualFile{
+			content: f0,
+			local:   files.NewReaderFile(f0.FileName(), path, f, stat),
+		}, nil
+	}
+}
+
+type dualFile struct {
+	content files.File
+	local   files.StatFile
+	buf     []byte
+}
+
+func (f *dualFile) IsDirectory() bool             { return false }
+func (f *dualFile) NextFile() (files.File, error) { return nil, files.ErrNotDirectory }
+func (f *dualFile) FileName() string              { return f.local.FileName() }
+func (f *dualFile) FullPath() string              { return f.local.FullPath() }
+func (f *dualFile) Stat() os.FileInfo             { return f.local.Stat() }
+func (f *dualFile) Size() (int64, error)          { return f.local.Stat().Size(), nil }
+
+func (f *dualFile) Read(res []byte) (int, error) {
+	// First read the content send from the client
+	n, err1 := f.content.Read(res)
+	if err1 == io.ErrUnexpectedEOF { // avoid this special case
+		err1 = io.EOF
+	}
+	if err1 != nil && err1 != io.EOF {
+		return 0, err1
+	}
+	res = res[:n]
+
+	// Next try to read the same amount of data from the local file
+	if n == 0 && err1 == io.EOF {
+		// Make sure we try to read at least one byte in order
+		// to get an EOF
+		n = 1
+	}
+	if cap(f.buf) < n {
+		f.buf = make([]byte, n)
+	} else {
+		f.buf = f.buf[:n]
+	}
+	n, err := io.ReadFull(f.local, f.buf)
+	if err == io.ErrUnexpectedEOF { // avoid this special case
+		err = io.EOF
+	}
+	if err != nil && err != io.EOF {
+		return 0, err
+	}
+	f.buf = f.buf[:n]
+
+	// Now compare the results and return an error if the contents
+	// sent from the client differ from the contents of the file
+	if len(res) == 0 && err1 == io.EOF {
+		if len(f.buf) == 0 && err == io.EOF {
+			return 0, io.EOF
+		} else {
+			return 0, errors.New("server side file is larger")
+		}
+	}
+	if !bytes.Equal(res, f.buf) {
+		return 0, errors.New("file contents differ")
+	}
+	return n, err1
+}
+
+func (f *dualFile) Close() error {
+	err := f.content.Close()
+	if err != nil {
+		return err
+	}
+	return f.local.Close()
+}
+
+var lsFileStore = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "List objects in filestore",
+		ShortDescription: `
```

Comment:
We have synopsis auto-generator in the pipeline (https://github.com/ipfs/go-ipfs/pull/2785) so I don't think it is needed.